File: backend/app/models/task.py
from pydantic import BaseModel, model_validator
from typing import Optional, Literal


class TaskTemplateCreate(BaseModel):
    title: str
    description: Optional[str] = None
    task_type: Literal["maintenance", "cleaning", "inspection"]
    property_id: str
    unit_id: Optional[str] = None
    assigned_to: str
    priority: str = "normal"
    recurrence_type: Literal["once", "daily", "weekly", "monthly", "quarterly"]
    recurrence_day_of_week: Optional[int] = None  # 0-6 (Monday=0)
    recurrence_day_of_month: Optional[int] = None  # 1-31
    recurrence_month_of_quarter: Optional[int] = None  # 1-3
    recurrence_start_date: str
    recurrence_end_date: Optional[str] = None

    @model_validator(mode="after")
    def validate_recurrence_fields(self):
        if self.recurrence_type == "weekly" and self.recurrence_day_of_week is None:
            raise ValueError("recurrence_day_of_week is required for weekly recurrence")
        if self.recurrence_type == "monthly" and self.recurrence_day_of_month is None:
            raise ValueError("recurrence_day_of_month is required for monthly recurrence")
        if self.recurrence_type == "quarterly":
            if self.recurrence_day_of_month is None:
                raise ValueError("recurrence_day_of_month is required for quarterly recurrence")
            if self.recurrence_month_of_quarter is None:
                raise ValueError("recurrence_month_of_quarter is required for quarterly recurrence")
        return self


class TaskTemplateUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    task_type: Optional[Literal["maintenance", "cleaning", "inspection"]] = None
    property_id: Optional[str] = None
    unit_id: Optional[str] = None
    assigned_to: Optional[str] = None
    priority: Optional[str] = None
    recurrence_type: Optional[Literal["once", "daily", "weekly", "monthly", "quarterly"]] = None
    recurrence_day_of_week: Optional[int] = None
    recurrence_day_of_month: Optional[int] = None
    recurrence_month_of_quarter: Optional[int] = None
    recurrence_start_date: Optional[str] = None
    recurrence_end_date: Optional[str] = None


class TaskTemplateResponse(BaseModel):
    id: str
    title: str
    description: Optional[str] = None
    task_type: str
    property_id: str
    property_name: Optional[str] = None
    unit_id: Optional[str] = None
    assigned_to: str
    assigned_to_name: Optional[str] = None
    priority: str = "normal"
    recurrence_type: str
    recurrence_day_of_week: Optional[int] = None
    recurrence_day_of_month: Optional[int] = None
    recurrence_month_of_quarter: Optional[int] = None
    recurrence_start_date: Optional[str] = None
    recurrence_end_date: Optional[str] = None
    is_active: bool = True
    next_occurrences: list[str] = []
    created_at: Optional[str] = None


class TaskUpdate(BaseModel):
    status: Optional[Literal["pending", "in_progress", "done", "cancelled"]] = None
    completion_notes: Optional[str] = None
    completion_photos: Optional[list] = None


class TaskResponse(BaseModel):
    id: str
    title: str
    description: Optional[str] = None
    task_type: str
    property_name: Optional[str] = None
    unit_number: Optional[str] = None
    assigned_to_name: Optional[str] = None
    assigned_by_name: Optional[str] = None
    priority: str = "normal"
    status: str = "pending"
    due_date: Optional[str] = None
    completed_at: Optional[str] = None
    created_at: Optional[str] = None


class CalendarTaskResponse(BaseModel):
    id: str
    title: str
    status: str
    priority: str
    due_date: Optional[str] = None
    property_name: Optional[str] = None
    assigned_to_name: Optional[str] = None

File: frontend/src/types/task.ts
export interface TaskTemplate {
  id: string;
  title: string;
  description?: string;
  task_type: string;
  property_id: string;
  property_name: string;
  unit_id?: string;
  assigned_to: string;
  assigned_to_name: string;
  priority: string;
  recurrence_type: 'once' | 'daily' | 'weekly' | 'monthly' | 'quarterly';
  recurrence_day_of_week?: number;
  recurrence_day_of_month?: number;
  recurrence_month_of_quarter?: number;
  recurrence_start_date: string;
  recurrence_end_date?: string;
  is_active: boolean;
  next_occurrences: string[];
  created_at: string;
}

export interface TaskTemplateCreate {
  title: string;
  description?: string;
  task_type: 'maintenance' | 'cleaning' | 'inspection';
  property_id: string;
  unit_id?: string;
  assigned_to: string;
  priority?: string;
  recurrence_type: 'once' | 'daily' | 'weekly' | 'monthly' | 'quarterly';
  recurrence_day_of_week?: number;
  recurrence_day_of_month?: number;
  recurrence_month_of_quarter?: number;
  recurrence_start_date: string;
  recurrence_end_date?: string;
}

export interface TaskTemplateUpdate {
  title?: string;
  description?: string;
  task_type?: string;
  property_id?: string;
  unit_id?: string;
  assigned_to?: string;
  priority?: string;
  recurrence_type?: 'once' | 'daily' | 'weekly' | 'monthly' | 'quarterly';
  recurrence_day_of_week?: number;
  recurrence_day_of_month?: number;
  recurrence_month_of_quarter?: number;
  recurrence_start_date?: string;
  recurrence_end_date?: string;
}

export interface Task {
  id: string;
  title: string;
  description?: string;
  task_type: string;
  property_name: string;
  unit_number?: string;
  assigned_to_name: string;
  assigned_by_name?: string;
  priority: 'low' | 'normal' | 'high' | 'urgent';
  status: 'pending' | 'in_progress' | 'done' | 'cancelled';
  due_date?: string;
  completed_at?: string;
  created_at: string;
}

export interface TaskUpdate {
  status?: 'pending' | 'in_progress' | 'done' | 'cancelled';
  completion_notes?: string;
  completion_photos?: string[];
}

export interface CalendarTask {
  id: string;
  title: string;
  status: string;
  priority: string;
  due_date?: string;
  property_name?: string;
  assigned_to_name?: string;
}

export interface StaffMember {
  id: string;
  full_name: string;
  email: string;
  role: string;
  phone?: string;
  is_active: boolean;
  assigned_properties: string[];
  notes?: string;
  created_at: string;
}

export interface StaffInvite {
  email: string;
  role: 'manager' | 'accountant' | 'maintenance' | 'cleaner';
  property_id?: string;
}

export interface StaffRoleUpdate {
  role: 'manager' | 'accountant' | 'maintenance' | 'cleaner';
}
